tiled/client/context.py

```diff
@@ -934,40 +854,60 @@ def _can_prompt():
-    # injects TILED_CACHE_DIR env var to use a temporary directory.
-    TILED_CACHE_DIR = Path(
-        os.getenv("TILED_CACHE_DIR", platformdirs.user_cache_dir("tiled"))
-    )
-    return Path(
-        TILED_CACHE_DIR, "default_identities", urllib.parse.quote_plus(str(api_uri))
-    )
-
-
-def set_default_identity(api_uri, provider, username):
-    """
-    Stash the identity used with this API so that we can reuse it by default.
-    """
-    filepath = _default_identity_filepath(api_uri)
-    filepath.parent.mkdir(parents=True, exist_ok=True)
-    with open(filepath, "w") as file:
-        json.dump({"username": username, "provider": provider}, file)
+def password_grant(http_client, auth_endpoint, provider, username, password):
+    form_data = {
+        "grant_type": "password",
+        "username": username,
+        "password": password,
+    }
+    token_response = http_client.post(auth_endpoint, data=form_data, auth=None)
+    handle_error(token_response)
+    return token_response.json()
+
+
+def device_code_grant(http_client, auth_endpoint):
+    verification_response = http_client.post(auth_endpoint, json={}, auth=None)
+    handle_error(verification_response)
+    verification = verification_response.json()
+    authorization_uri = verification["authorization_uri"]
+    print(
+        f"""
+You have {int(verification['expires_in']) // 60} minutes visit this URL
 
+{authorization_uri}
 
-def get_default_identity(api_uri):
-    """
-    Look up the default identity to use with this API.
-    """
-    filepath = _default_identity_filepath(api_uri)
-    if filepath.exists():
-        return json.loads(filepath.read_text())
+and enter the code:
 
+{verification['user_code']}
 
-def clear_default_identity(api_uri):
-    """
-    Clear the cached default identity for this API.
-    """
-    filepath = _default_identity_filepath(api_uri)
-    if filepath.exists():
-        filepath.unlink()
+"""
+    )
+    import webbrowser
+
+    webbrowser.open(authorization_uri)
+    deadline = verification["expires_in"] + time.monotonic()
+    print("Waiting...", end="", flush=True)
+    while True:
+        time.sleep(verification["interval"])
+        if time.monotonic() > deadline:
+            raise Exception("Deadline expired.")
+        # Intentionally do not wrap this in handle_error(...).
+        # Check status codes manually below.
+        access_response = http_client.post(
+            verification["verification_uri"],
+            json={
+                "device_code": verification["device_code"],
+                "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
+            },
+            auth=None,
+        )
+        if (access_response.status_code == httpx.codes.BAD_REQUEST) and (
+            access_response.json()["detail"]["error"] == "authorization_pending"
+        ):
+            print(".", end="", flush=True)
+            continue
+        handle_error(access_response)
+        print("")
```

Comment:
Should it display a "success" message here, or does the equivalent happen in an outer function?

```suggestion
        print("\nSUCCESS: Authenticated")
```

Reply:
The equivalent happens in the caller (`prompt_for_credentials`) where we print a message that indicates not only "success" but specifically the username that you have just successfully logged in with. The exact message is customizable by the server, so we can comply with rules like the ORCID Terms of Service (which requires a confirmation that includes the ORCID digits) or BNL Cyber rules (which require the boilerplate about accessing a Federal system).